docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,810 @@
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+- Generators create a pipeline, starting from base input sources and mapping them to the output they wish to produce. The more exposed,
+  properly equatable states exist, the earlier the compiler will be able to cut off changes and reuse the same output.
+
+## Out of scope designs
+
+We will briefly look at the non-solvable problems as examples of the kind of problems source generators are *not* designed to solve:
+
+### Language features
+
+Source generators are not designed to replace new language features: for instance one could imagine [records](records.md) being implemented as a source generator
+that converts the specified syntax to a compilable C# representation.
+
+We explicitly consider this to be an anti-pattern; the language will continue to evolve and add new features, and we don't expect source generators to be
+a way to enable this. Doing so would create new 'dialects' of C# that are incompatible with the compiler without generators. Further, because generators,
+by design, cannot interact with each other, language features implemented in this way would quickly become incompatible with other additions to the language.
+
+### Code rewriting
+
+There are many post-processing tasks that users perform on their assemblies today, which here we define broadly as 'code rewriting'. These include, but
+are not limited to:
+
+- Optimization
+- Logging injection
+- IL Weaving
+- Call site re-writing
+
+While these techniques have many valuable use cases, they do not fit into the idea of *source generation*. They are, by definition, code altering operations
+which are explicitly ruled out by the source generators proposal.
+
+There are already well supported tools and techniques for achieving these kinds of operations, and the source generators proposal is not aimed at replacing them.
+We are exploring approaches for call site rewriting (see [interceptors.md](interceptors.md)), but those features are experimental and may change significantly
+or even be removed.
+
+## Conventions
+
+TODO: List a set of general conventions that apply to all designs below. E.g. Re-using namespaces, generated file names etc.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to add a type to the compilation, that can be referenced by the user's code. Common use cases include
+creating an attribute that will be used to drive other source generator steps.
+
+**Solution:** Have the user write the code as if the type was already present. Generate the missing type based on information available in the compilation using
+the `RegisterPostInitializationOutput` step.
+
+**Example:**
+
+Given the following user code:
+
+```csharp
+public partial class UserClass
+{
+    [GeneratedNamespace.GeneratedAttribute]
```

Comment:
I guess I don't understand this feedback? It _is_ produced by the generator; that's actually what I was going for?

Reply:
Hmm ... that seems a bit odd to me. Generating the attribute every time leads to other warnings once you ref DLLs that used the same generator as you. I would assume we recommended that the DLL paired with the generator have the attribute. Runtime is weird here but would assume that's the preferred setup for most other authors.

If we want to use this as the recommended pattern then we probably need to revisit the CS0436 diagnostic so it's suppressible by the generator itself.

Reply:
FWIW, what's relatively common practice for this specific case (I'm doing that [in one of my libraries](https://github.com/Sergio0694/privatelike/blob/main/src/PrivateLike/PrivateLikeAttributeDiagnosticSuppressor.cs) and I know of at least [another project doing the same](https://github.com/k94ll13nn3/AutoConstructor/blob/dcb264dae15c7319e0286a2a7398f583af31617c/src/AutoConstructor.Generator/Suppressors/AutoConstructorAttributeDiagnosticSuppressor.cs#L8)) is to bundle a diagnostic suppressor that can automatically suppress the warning when it's being produced by the attribute type it knows is being generated by the same generator (it can just match by name). Though personally I think it would maybe make more sense to say "bundle a runtime library for the attributes, and a generator in the same NuGet package" as the recommended approach. Generating attributes on the fly like this is not something I think most libraries should do, except for very specific scenarios where you absolutely do not want to have any runtime library at all (eg. that's the case for [DNNE](https://github.com/AaronRobinsonMSFT/DNNE)).